crud/readview.lua

```diff
@@ -0,0 +1,613 @@
+local errors = require('errors')
+
+local fiber = require('fiber')
+local common = require('crud.select.compat.common')
+
+local Merger = require('crud.select.merger')
+
+local stash = require('crud.common.stash')
+local utils = require('crud.common.utils')
+local sharding = require('crud.common.sharding')
+local select_executor = require('crud.select.executor')
+local select_filters = require('crud.compare.filters')
+local dev_checks = require('crud.common.dev_checks')
+local schema = require('crud.common.schema')
+local const = require('crud.common.const')
+local sharding_metadata_module = require('crud.common.sharding.sharding_metadata')
+local stats = require('crud.stats')
+local checks = require('checks')
+
+local compare_conditions = require('crud.compare.conditions')
+local select_plan = require('crud.compare.plan')
+
+local ReadviewError = errors.new_class('ReadviewError', {capture_stack = false})
+
+local readview = {}
+
+function checkers.vshard_call_mode(p)
+    return p == 'write' or p == 'read'
+end
+
+local function readview_open_on_storage(readview_name)
+    local read_view = box.read_view.open({name = readview_name})
+
+    if read_view == nil then
+        return "error creating readview"
+    end
+
+    local result = {}
+    result.uuid = box.info().uuid
+    result.id = read_view.id
+
+    return result
+end
+
+local function readview_close_on_storage(readview_uuid)
+    dev_checks('table')
+
+    local list = box.read_view.list()
+
+    local readview_id
+    for _, replica_info in pairs(readview_uuid) do
+        if replica_info.uuid == box.info().uuid then
+            readview_id = replica_info.id
+        end
+    end
+
+    if readview_id == nil then
+        return ""
+    end
+
+    for k,v in pairs(list) do
+        if v.id == readview_id then
+            list[k]:close()
+            return list[k]:info()
+        end
+    end
+
+    return "no such readview"
+end
+
+local function select_readview_on_storage(space_name, index_id, conditions, opts, readview_name)
+    dev_checks('string', 'number', '?table', {
+        scan_value = 'table',
+        after_tuple = '?table',
+        tarantool_iter = 'number',
+        limit = 'number',
+        scan_condition_num = '?number',
+        field_names = '?table',
+        sharding_key_hash = '?number',
+        sharding_func_hash = '?number',
+        skip_sharding_hash_check = '?boolean',
+        yield_every = '?number',
+        fetch_latest_metadata = '?boolean',
+    }, 'number')
```

Comment:
it seems that it is `readview_id` now. You can use as an additional option.

Reply:
changed